### DISCUSSION TITLE
SSR using vitejs

### DISCUSSION THREAD

**Discussion by ibrunotome** [inertiajs/inertia#1038] [Help (Vue)] [answered]
I'm using inertia with vitejs instead webpack, everything works well but without any settings related to SSR (see below):

```js
const { resolve } = require('path')
import vue from '@vitejs/plugin-vue'

export default ({ command }) => ({
  base: command === 'serve' ? '' : '/dist/',
  publicDir: 'fake_dir_so_nothing_gets_copied',
  build: {
    manifest: true,
    outDir: resolve(__dirname, 'public/dist'),
    rollupOptions: {
      input: 'resources/scripts/app.ts',
      external: [
        resolve(__dirname, 'public/img')
      ],
    },
  },

  plugins: [vue()],

  resolve: {
    alias: {
      '@': resolve('./resources/scripts'),
    },
  },
})
```

Has anyone made it with SSR?

**Comment by ajnsn** [comment, marked as answer]
Hey @ibrunotome 

I am also still trying out things regarding SSR in combination with Vite, but this seems to work:

**vite.ssr.config.js**

```js
import { defineConfig } from 'vite';
import { resolve } from 'path';
import vue from '@vitejs/plugin-vue';

export default defineConfig(() => ({
    publicDir: false,
    build: {
        ssr: true,
        target: 'node17',
        outDir: 'public/build-ssr',
        rollupOptions: {
            input: 'resources/js/ssr.js',
        },
    },
    resolve: {
        alias: {
            '@': resolve(__dirname, 'resources/js'),
        },
    },
    plugins: [vue()],
}));
```

**ssr.js**

```js
import { createSSRApp, h } from 'vue';
import { renderToString } from '@vue/server-renderer';
import { createInertiaApp } from '@inertiajs/inertia-vue3';
import createServer from '@inertiajs/server';

const components = import.meta.globEager('./pages/**/*.vue');

createServer(page =>
    createInertiaApp({
        page,
        render: renderToString,
        resolve: name => components[`./pages/${name}.vue`].default,
        setup({ app, props, plugin }) {
            return createSSRApp({
                render: () => h(app, props),
            }).use(plugin);
        },
    }),
);
```

```
vite build --config vite.ssr.config.js
node public/build-ssr/ssr.js
```

**Reply by ibrunotome** [reply]
Wow! Thank you for that! @ajsn

Now it's time to handle some access of `document` by a i18n lib and this error on my side: `[vite:vue] Custom directive is missing corresponding SSR transform and will be ignored.` for directives like v-mask or v-touch 💪

[edit] directives fixed with https://github.com/vueuse/patch-vue-directive-ssr

**Reply by RobertBoes** [reply]
This page might help you with that: https://ssr.vuejs.org/guide/universal.html#data-reactivity-on-the-server
Especially the parts about using hooks to access `document`/`window` and, if possible, providing a "server-side version" of your custom directives

**Reply by ajnsn** [reply]
What you can also do about the directives is to transform them inside the Vue plugin config, e.g.

```js
const ssrTransform = () => {
    return {
        props: [],
        needRuntime: true,
    };
};

export default defineConfig(() => ({
    ...
    plugins: [
        vue({
            template: {
                ssr: true,
                compilerOptions: {
                    directiveTransforms: {
                        mask: ssrTransform,
                        touch: ssrTransform,
                    },
                },
            },
        }),
    ],
}));
```

At least for now I also  do not have any better solution for this.

**Reply by secondmanveran** [reply]
I got everything to run locally ... though to be honest I don't know what I'm looking for. I just kept at it until the SSR server started and didn't produce any errors when I nav to a new page.🤷🏼‍♂️

I thought that I would do a view page source and see the content that comes from that page ... and I don't ... I do see some content on a couple pages, terms and privacy ... this is a Laravel Jetstream app with Inertia.

The biggest issue so far is that it won't run after a CI deployment.

I have Github Actions do a release if needed, create a new branch from the latest tag, compile the app and all the assets via vite, including the new ssr.js server which is rendered to `public/render/ssr.js` then the files are added and committed, and deployed to forge, and lastly the newly created release branch is removed.

I use this same workflow on everything I build.

When I try to launch the ssr server `node public/render/ssr.js` I get an error that it cannot find Vue.

Why would Vue not be found in the SSR file? You can't possibly tell me that I need a `node_modules` folder installed on my server ... that would be the most counter productive thing imaginable. Why even use a CI process ...

This is the full error:

```bash
internal/modules/cjs/loader.js:905
  throw err;
  ^

Error: Cannot find module 'vue'
Require stack:
- /home/forge/example.com/public/render/ssr.js
  at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
  at Function.Module._load (internal/modules/cjs/loader.js:746:27)
  at Module.require (internal/modules/cjs/loader.js:974:19)
  at require (internal/modules/cjs/helpers.js:93:18)
  at Object.<anonymous> (/home/forge/example.com/public/render/ssr.js:21:11)
  at Module._compile (internal/modules/cjs/loader.js:1085:14)
  at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
  at Module.load (internal/modules/cjs/loader.js:950:32)
  at Function.Module._load (internal/modules/cjs/loader.js:790:12)
  at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12) {
code: 'MODULE_NOT_FOUND',
requireStack: [ '/home/forge/example.com/public/render/ssr.js' ]
}
```

Any ideas?

**Reply by ibrunotome** [reply]
@ajnsn after two weeks working on other tasks, I'm finishing SSR settings again. The build is working, but when I hit a url that has the Ziggy `route` method, it crashes. Works normally on pages that doesn't has the `route` method.

<img width="652" alt="Screen Shot 2022-01-29 at 19 39 41" src="https://user-images.githubusercontent.com/4256471/151679852-9be524f5-50d2-4ada-8b27-49d8d49c00d3.png">

This is my route method: 
<img width="809" alt="Screen Shot 2022-01-29 at 19 42 20" src="https://user-images.githubusercontent.com/4256471/151679901-0086b992-c6a0-4578-8d1d-c7a4ae7059cc.png">

I use it normally with `import route from '@/Shared/useRoute'` on my .vue files and it works normally without SSR. 
<img width="426" alt="Screen Shot 2022-01-29 at 19 48 46" src="https://user-images.githubusercontent.com/4256471/151680064-268b827c-61e7-4b96-b838-d225e62d86b0.png">

Do you have a idea of what could be happening?

**Reply by ajnsn** [reply]
Hey @ibrunotome 
afaik Ziggy heavily relies on the `window`, which is not available in SSR. 
[This article](https://aaronfrancis.com/2022/using-ziggy-with-inertia-server-side-rendering) might help. Also make sure to use the latest version(s).

**Reply by ibrunotome** [reply]
> Hey @ibrunotome
> afaik Ziggy heavily relies on the `window`, which is not available in SSR.
> [This article](https://aaronfrancis.com/2022/using-ziggy-with-inertia-server-side-rendering) might help. Also make sure to use the latest version(s).

I came to this article yesterday, but since I'm using Composition API and typescript, the suggested mixin method on the article would not work since Composition API doesn't have mixins. So I edited my useRoute.ts to replace the Ziggy configuration like the article showed, but only when the request is coming from a server and... it worked 😀

Thank you for your help Alexander!

**Reply by secondmanveran** [reply]
@ibrunotome 

Can you show the new `useRoute.ts` I'm interested in seeing how you resolved this?

Thanks

**Reply by ibrunotome** [reply]
@secondmanveran 

```ts
import ZiggyRoute, { Config, RouteParamsWithQueryOverload } from 'ziggy-js'
import { Ziggy } from '@/Shared/useZiggy'
import { computed } from 'vue'
import { usePage } from '@inertiajs/inertia-vue3'

const locale = computed(() => usePage().props.value.locale)

// @ts-ignore
const ZiggyConfig = computed<Config>(() => ({
  ...Ziggy,
  // @ts-ignore
  location: new URL(usePage().props.value.ziggy.location),
}))

const route = (name, params?, absolute = false, config = ZiggyConfig.value) =>
  ZiggyRoute(name, params, absolute, config)

export const localizedRoute = (routeName: string, params?: RouteParamsWithQueryOverload) => {
  if (locale.value === 'en') {
    return route(routeName, params)
  }

  return route(`${locale.value}.${routeName}`, params)
}

export default route
```
